### PR TITLE
Bump dpy to v2, and update the project to use BotBase

### PR DESCRIPTION
**Pull request by D0rs4n**

### CODE REVIEW

File: bot/__main__.py [hunk 1]
```diff
@@ -21,25 +44,23 @@ async def main() -> None:
...
+                redis_session=await _create_redis_session(),
+                http_session=session,
+                guild_id=constants.Client.guild,
+                allowed_roles=allowed_roles,
+                command_prefix=constants.Client.prefix,
+                activity=discord.Game("The Not-Quite-So-Bot-as-Sir-Lancebot"),
+                intents=_intents,
+            )
+            async with bot.instance:
+                bot.instance._guild_available = asyncio.Event()
```
Comment (ChrisLovering) [inline bot/__main__.py:63]: Why is this needed here? It gets set to this in BotBase.setup_hook

**Review by ChrisLovering — CHANGES_REQUESTED** [review]
`on_guild_available`, `on_guild_unavailable` and `wait_until_guild_available` can all be removed, as they are implemented in BotBase.

Also we could move the init of `self.code_jam_mgmt_api` to the setup_hook, that way we don't need to have a login method in SirRobin.

File: bot/__main__.py (hunk 1 above)
Comment (D0rs4n) [reply bot/__main__.py]: Right, sorry. Patched in 443c2c9

**Review by ChrisLovering — APPROVED** [review]
Looks good, thanks for this :)

File: bot/bot.py [hunk 2]
```diff
@@ -1,62 +1,38 @@
...
-        # This session may want to be recreated on login/disconnect.
-        # Additionally, on the bot repo we use a different connector that is more "stable".
-        self.code_jam_mgmt_api: Optional[APIClient] = None
-        self.http_session: Optional[aiohttp.ClientSession] = None
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
 
-        self._guild_available: Optional[asyncio.Event] = None
-        self.redis_session = redis_session
+        self.code_jam_mgmt_api: APIClient = None
```
Comment (ionite34) [inline bot/bot.py:19]:
```suggestion
        self.code_jam_mgmt_api: APIClient | None = None
```

The change from `Optional[APIClient]` to `APIClient` would appear to make this type hint incorrect, as it's being initialized with`None`

**Review by ionite34 — CHANGES_REQUESTED** [review]

**Review by ionite34 — APPROVED** [review]
All looks good
